cfme/tests/intelligence/chargeback/test_resource_allocation.py

```diff
@@ -0,0 +1,296 @@
+
+    memory_alloc_cost = resource_cost(appliance, provider, 'Allocated Memory',
+        memory_alloc, description, 'Compute')
+
+    vcpu_alloc_cost = resource_cost(appliance, provider, 'Allocated CPU Count',
+        vcpu_alloc, description, 'Compute')
+
+    return {"storage_alloc_cost": storage_alloc_cost,
+            "memory_alloc_cost": memory_alloc_cost,
+            "vcpu_alloc_cost": vcpu_alloc_cost}
+
+
+@pytest.yield_fixture(scope="module")
+def chargeback_report_custom(appliance, vm_ownership, assign_custom_rate, provider):
+    """Create a Chargeback report based on a custom rate; Queue the report"""
+    owner = vm_ownership
+    data = {
+        'menu_name': 'cb_custom_{}'.format(provider.name),
+        'title': 'cb_custom_{}'.format(provider.name),
+        'base_report_on': 'Chargeback for Vms',
+        'report_fields': ['Memory Allocated Cost', 'Memory Allocated over Time Period', 'Owner',
+        'vCPUs Allocated over Time Period', 'vCPUs Allocated Cost',
+        'Storage Allocated', 'Storage Allocated Cost'],
+        'filter': {
+            'filter_show_costs': 'Owner',
+            'filter_owner': owner,
+            'interval_end': 'Today (partial)'
+        }
+    }
+    report = appliance.collections.reports.create(is_candu=True, **data)
+
+    logger.info('Queuing chargeback report with custom rate for {} provider'.format(provider.name))
+    report.queue(wait_for_finish=True)
+
+    if not list(report.saved_reports.all()[0].data.rows):
+        pytest.skip('Empty report')
+    else:
+        yield list(report.saved_reports.all()[0].data.rows)
+
+    if report.exists:
+        report.delete()
+
+
+@pytest.yield_fixture(scope="module")
+def new_chargeback_rate():
+    """Create a new chargeback rate"""
+    desc = 'custom_{}'.format(fauxfactory.gen_alphanumeric())
+    compute = rates.ComputeRate(description=desc,
+                fields={'Allocated CPU Count':
+                        {'per_time': 'Hourly', 'variable_rate': '2'},
+                        'Allocated Memory':
+                        {'per_time': 'Hourly', 'variable_rate': '2'}})
+    compute.create()
+    if not BZ(1532368, forced_streams=['5.9']).blocks:
+        storage = rates.StorageRate(description=desc,
+                fields={'Allocated Disk Storage':
+                        {'per_time': 'Hourly', 'variable_rate': '3'}})
+    storage.create()
+    yield desc
+
+    if compute.exists:
+        compute.delete()
+    if not BZ(1532368, forced_streams=['5.9']).blocks and storage.exists:
+        storage.delete()
+
+
+@pytest.mark.parametrize('resource_cost',
+                         ['memory', 'vcpu', 'storage'],
+                         ids=['memory_alloc', 'vcpu_alloc', 'storage_alloc'])
+def test_validate_cost(chargeback_costs_custom, chargeback_report_custom, soft_assert,
+        resource_cost):
+
+    for groups in chargeback_report_custom:
+        if not (groups['Memory Allocated Cost'] or groups['Storage Allocated Cost'] or
```

Comment:
Question: Shouldn't this test fail if we don't get some of the columns?

Reply:
@rlbabyuk , thanks for reviewing the PR.

No, the tests are for validating costs in a report and I have implemented the tests such that the tests only fail if the costs don't match.

If the columns are missing, it's more of a report issue .